tidb-lightning/tidb-lightning-error-resolution.md

```diff
@@ -0,0 +1,327 @@
+
+This article describes how to use use the type error feature (`lightning.max-error`) and the duplicate resolution feature (`tikv-importer.duplicate-resolution`). It also introduces the database where these errors are stored (`lightning.task-info-schema-name`). At the end of this article, it provides an example.
+
+## Type error
+
+> **Warning:**
+>
+> The TiDB Lightning type error (`lightning.max-error`) is an experimental feature. It is **NOT** recommended to only rely on it in the production environment.
+
+You can use the configuration `lightning.max-error` to increase the tolerance of errors related to data types. If this is set to *N*, TiDB Lightning will allow and skip up to *N* errors from the data source before aborting. The default value 0 means any single error is fatal.
+
+These errors are recorded in a database. After the import is completed, you can view the errors in the database and process them manually. For more information, see [Error Report](#Error-report).
+
+{{< copyable "" >}}
+
+```toml
+[lightning]
+max-error = 0
+```
+
+The following errors are governed by this configuration:
+
+* Invalid value (example: set `'Text'` to an INT column).
+* Numeric overflow (example: set 500 to a TINYINT column)
+* String overflow (example: set `'Very Long Text'` to a VARCHAR(5) column).
+* Zero date time (namely `'0000-00-00'` and `'2021-12-00'`).
+* Set NULL to a NOT NULL column.
+* Failed to evaluate a generated column expression.
+* Column count mismatch. Number of values in the row does not match number of columns of the table.
+* Unique/Primary key conflict in TiDB backend, when `on-duplicate = "error"`.
+* Any other SQL errors.
+
+The following errors are always fatal, and cannot be skipped by changing `max-error`:
+
+* Syntax error (such as unclosed quotation marks) in the original CSV, SQL or Parquet file.
+* I/O, network or system permission errors.
+
+Unique/Primary key conflict in Local backend is handled separately, explained in the next section.
+
+## Duplicate resolution in Local-backend mode
+
+Local-backend mode imports data by first converting them to KV pairs, and ingest them into TiKV in batches. Unlike the TiDB-backend mode, duplicate rows are not detected until the end of task. Therefore, duplicate errors in Local-backend mode is not controlled by `max-error`, but rather a separate configuration `duplicate-resolution`.
+
+{{< copyable "" >}}
+
+```toml
+[tikv-importer]
+duplicate-resolution = 'record'
+```
+
+There are three possible values of `duplicate-resolution`:
+
+* **'none'** — Do not detect duplicates. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
+* **'record'** — Detect duplicates, but do not attempt to fix it. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
```

Comment:
@gozssky Please confirm this part

Reply:
`will fail checksum check` -> `will skip checksum and report conflict errors count`